src/ranged.cpp

```diff
@@ -830,13 +850,72 @@ static void do_aim( player *p, std::vector <Creature *> &t, int &target,
     }
 }
 
+static int print_aim_bars( const player &p, WINDOW *w, int line_number, item *weapon,
+                           Creature *target, int predicted_recoil ) {
+    // This is absolute accuracy for the player.
+    // TODO: push the calculations duplicated from Creature::deal_projectile_attack() and
+    // Creature::projectile_attack() into shared methods.
+    // Dodge is intentionally not accounted for.
+
+    // Confidence is chance of the actual shot being under the target threshold,
+    // This simplifies the calculation greatly, that's intentional.
+    const double aim_level = predicted_recoil + p.driving_recoil +
+        p.get_weapon_dispersion( weapon, false );
+    const double range = rl_dist( p.pos(), target->pos() );
+    const double missed_by = aim_level * 0.00021666666666666666 * range;
+    const double hit_rating = missed_by / std::max( double( p.get_speed() ) / 80., 1.0 );
```

Comment:
Not related to the refactor, but why does this formula involve speed? And speed of the shooter at that.